### ISSUE TITLE
SOGS Session ID blinding

### ISSUE THREAD

**Issue by jagerman** [oxen-io/session-pysogs#28]
Brainstorming:

0. Alice, Bob, and Chuck are in a SOGS (the DNA room which is all about the number 42, of course, but recommends that you bring a towel).
1. In order to participate in the room Alice derives a shadow private and public key according to:
    ```
    k = H(open group pubkey)
    a' = ka
    A' = kA
    B = 0x04 || A'
    ```
    and submits SOGS messages using a per-server, pseudo-session-id `B` (which is like a Session ID but, in hex notation, begins with 04 for differentiation).
2. Alice submits a post to the DNA room, signing the message using `a'` (using the same XEd25519 signatures as are used currently, but with this derived key instead), and uses `A'` as the publicly visible sender session ID.
3. Bob retrieves messages from SOGS and receives the message from Alice.  Because Bob and Alice have prior direct contact, his session client can derive `A'`, and so Bob's client automatically displays the message as being from Alice (and shows both derived and real Session IDs when viewing message details).
4. Chuck, who does not know Alice's true session ID, receives the same message, can verify the signature, but since `A'` doesn't match the derived keys of any of Chuck's contacts, can only display the message using the pseudo-id (the one starting with 04).
5. Chuck wants to DM Alice, but in order to do so requires Alice's Session ID.
6. Chuck constructs a "please send your session ID to me" message (precise encoding TBD), which includes Chuck's own Session ID, encrypted using Alice's pseudo-id, and sends it to the SOGS for delivery to Alice.
7. Alice receives it from SOGS the next time she polls for messages, decrypts it, and then, if she chooses to allow it (perhaps automatically or via confirmation depending on Privacy settings) constructs a "here's my session ID" message that goes directly to Chuck's session ID, which Chuck then recognizes as part of the DM he is trying to initiate.
8. Chuck receives, verifies, and opens a new conversation with Alice.  (And going forward, like Bob, can see when Alice posts a message).

Notes:
- There is some trust in the server during the discovery process.  A malicious server *could* fully intercept all messages, rewrite them using new IDs and passively translate IDs back and forth to keep itself in the middle of all conversations.  This is outside the scope of problems we can solve because it is the exact same problem as with *any* session ID discovery process.  It is, however, important that we include some identifying characteristics for the server in question so that the user can make an informed decision as to whether to engage in the handshake based on the server source.
- Although this proposal avoids linking unknown users based on their session ID (because each user will have a different psuedo-ID on each server) there are still various ways to link users:
  - ONS users are not protected if their ONS name matches their profile name (or is otherwise advertised): Session clients can instantly look up the ONS name and verify that the pseudo-id matches the published name.  (This is probably a *good* thing in some respects -- ONS users can get a special badge).
  - avatar uploads can link a user as having the same file server avatar URL in two different profile sections on two different servers almost certainly means you are the same user.  (This doesn't help you get the Session ID, but does allow linking of accounts).  Future capabilities might involve some sort of "don't post my avatar in messages to this server", or per-server avatars/display names.

**Comment by majestrate** [comment]
this was fixed with v0.3.0 tag right?

**Comment by KeeJef** [comment]
> this was fixed with v0.3.0 tag right?

Yes, Closing as this is now implemented
